### PR TITLE
Fix some incorrect SpecialFolder entries for Unix

### LINKED ISSUE #40353: System.Environment.GetFolderPath (Environment.SpecialFolder.ApplicationData) Is the wrong path on MacOS
### Description

When you call the following code, it is correct on Windows and Linux, but MacOS you get the linux path
`var path = System.Environment.GetFolderPath (Environment.SpecialFolder.ApplicationData) `
paths actual value = `~/.config `
It should be` ~/Library/Application Support`
### Configuration

MacOS

### LINKED ISSUE #63214: SpecialFolder.MyDocuments / Personal returning wrong path for Unix
### Description

Using `Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments)` (and by extensions also `SpecialFolder.Personal`) on Unix based systems does not return the documents folder, and instead returns the same path as `Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)`.

### Reproduction Steps

```cs
Console.WriteLine("Environment.SpecialFolder.MyDocuments:");
Console.WriteLine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments));
Console.WriteLine("Environment.SpecialFolder.UserProfile:");
Console.WriteLine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile));
```

### Expected behavior

Since `Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments)` should return the directory that serves as a common repository for documents, it should report the directory I have set via `xdg-users-dir` / `XDG_DOCUMENTS_DIR` or `~/Documents` on Mac.

### Actual behavior

It reports my home directory instead.
Comparison between Linux and Windows:
grafik

Checking [this](https://github.com/dotnet/runtime/blob/af9202a6daf91ba58d5d58bafbb6581aebb8a642/src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs#L89) it seems that it's intentional. Why doesn't it just use `ReadXdgDirectory(home, "XDG_DOCUMENTS_DIR", "Documents");` there? I can do a PR if it's wished for.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

Dotnet version: 6.0.101
Can reproduce on the following distros: Arch Linux WSL2, Ubuntu 20.04 WSL2, Arch Linux, macOS Big Sur
Architecture: x64

### Other information

_No response_

### LINKED ISSUE #68323: Environment.GetFolderPath(Environment.SpecialFolder.MyVideos) returns empty string in OSX
### Description

It returns an empty string on macOS Monterey running under .NET core 5.
My Electron.NET app runs under .NET Core 5.
It used to work well on the previous versions of OSX, now it returns an empty string.
Environment.GetFolderPath(Environment.SpecialFolder.MyMusic) returns correct path.

### Reproduction Steps

Call the API.

### Expected behavior

Return "/Users/USERNAME/Movies"

### Actual behavior

Returns an empty string

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

- Microsoft.NETCore.App 5.0.16 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
- macOS Monterey 12.1
- osx.12-x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by Miepee**
This PR:
- fixes an incorrect Documents/Personal entry for Linux+Mac (from `HOME` to `XDG_DOCUMENTS_DIR`/`HOME/Documents` for Linux and Mac respectively)
- fixes an incorrect Videos entry for Mac (from `HOME/Videos` to `HOME/Movies`)
- fixes an incorrect (Local) ApplicationData entry for Mac (from `XDG_CONFIG_HOME` and `XDG_DATA_HOME` to `HOME/Library/Application Support` for both ApplicationData and LocalApplicationData)

- Changes the Unix tests to check that `SpecialFolder.Personal` and `SpecialFolder.MyDocuments` are the Documents directory instead of `HOME`.

Fixes #63214, 
Fixes #68323, 
Fixes #40353

### CODE REVIEW

**Comment by Miepee** [comment]
Hm, a bit confused on that failed Windows Arm build. Can't see anything I did that broke it, nor see from the logs on where it error'd on my changes.

**Comment by danmoseley** [comment]
It's not you that broke these tests. @MattGal  if this from helix?

running %HELIX_CORRELATION_PAYLOAD%\scripts\ce25ee46de8f44688fa71d0af86d03fb\execute.cmd in C:\h\w\BEED09AC\w\A71708E3\e max 900 seconds

Output:

Exit Code:-900

**Comment by MattGal** [comment]
> It's not you that broke these tests. @MattGal is this from helix?
> 
> running %HELIX_CORRELATION_PAYLOAD%\scripts\ce25ee46de8f44688fa71d0af86d03fb\execute.cmd in C:\h\w\BEED09AC\w\A71708E3\e max 900 seconds
> 
> Output:
> 
> Exit Code:-900

Yes we hotfix-rolled-back a change that was causing this last week (https://github.com/dotnet/runtime/issues/67728) .  If you're still seeing it today we should investigate but checking out similar runs I think we're OK.  We may need some variant of the fix back though, but based on the date that's all this is.

**Comment by Miepee** [comment]
Since [os-mac-os-x](https://github.com/dotnet/runtime/labels/os-mac-os-x) got labeled for this, someone might wanna label this with [os-linux](https://github.com/dotnet/runtime/labels/os-linux) as well.

**Comment by danmoseley** [comment]
To enumerate breaking changes, on Mac, AppData goes from `/home/.config` today to `/home/Library/Application Support` and LocalAppData goes from `/home/.local/share` to `/home/Library/` and Documents/Personal from `/home` to `/home/Documents` and Videos from `/home/Videos` to `/home/Movies` . While on Linux, Documents/Personal goes from `/home` to XDG_DOCUMENTS_DIR.

@marklio thoughts about the impact of the break here? As always, it's a major update, so if we think it's the right change, it may well be acceptable to correct all these.

@dotnet/area-system-runtime thoughts?

It's quite remarkable how much discussion this has involved, and this is only some of the proposed changes. https://github.com/dotnet/runtime/issues/63214 is just one of the long discussions.

My inclination is to take this and look for feedback on Preview 5+.

**Comment by danmoseley** [comment]
The code changes look fine to me though.

**Comment by danmoseley** [comment]
rerunning validation

**Comment by jkotas** [comment]
> thoughts about the impact of the break here? 

The mobile platforms are probably the most impacted here.

cc @marek-safar

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs [hunk 1]
```diff
@@ -85,31 +85,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
...
-
                 case SpecialFolder.Desktop:
                 case SpecialFolder.DesktopDirectory:
                     return ReadXdgDirectory(home, "XDG_DESKTOP_DIR", "Desktop");
                 case SpecialFolder.Templates:
                     return ReadXdgDirectory(home, "XDG_TEMPLATES_DIR", "Templates");
-                case SpecialFolder.MyVideos:
-                    return ReadXdgDirectory(home, "XDG_VIDEOS_DIR", "Videos");
-
 #if TARGET_OSX
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: Should this ifdef be enabled for all Apple platforms?

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: Good question. I sadly only have one Apple device which is running MacOS.
Don't the mobile platforms get the folder from a different place though? I thought they'd get it from [here](https://github.com/dotnet/runtime/blob/fa284882f1fec209d2cadd915c621078e8041228/src/mono/System.Private.CoreLib/src/System/Environment.iOS.cs#L35) or is this mono specific?

**Comment by danmoseley** [comment]
What Mono had, for comparison [here](https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.cs) and for iOS [here](https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.iOS.cs) 

We have code specific for iOS [here](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/mono/System.Private.CoreLib/src/System/Environment.iOS.cs#L35) and for Android [here](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/libraries/System.Private.CoreLib/src/System/Environment.Android.cs#L16)

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: It does [not apply for iOS](https://github.com/dotnet/runtime/blob/3535e0769f202ae4cd820bea24afd20cee313966/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L2176). Is that all other Apple platforms?

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: From quick skimming the RIDs, there's TVOS and MacCatalyst as well. MacCatalyst is inheriting from iOS, tvOS is only inheriting from unix. I cannot however find much documentation regarding tvOS' file system right now. Does seem relatively locked down though, with apps not meant to write to local storage and [mostly to the cloud instead](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppleTV_PG/iCloudStorage.html#//apple_ref/doc/uid/TP40015241-CH10-SW1).

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: > It does not apply for iOS

Ah ok, I have missed that this file is ifdefed out.

Is there a good reason for the default folder handling to differ between macOS and other Apple OSes? Should macOS be switched to just use https://github.com/mono/mono/blob/main/mcs/class/corlib/System/Environment.iOS.cs?

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs [hunk 2]
```diff
@@ -85,31 +85,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
...
+                case SpecialFolder.ApplicationData:
+                    return Path.Combine(home, "Library", "Application Support");
+                case SpecialFolder.LocalApplicationData:
+                    return Path.Combine(home, "Library");
```
Comment (tmds) [inline src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: Apps store their settings in `ApplicationData` and `LocalApplicationData`. Changing the locations will cause apps to no longer find their settings.

With no access to the previous location, implementing setting migration is also difficult.

Besides it being more correct, is there a tangible improvement?

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs [hunk 3]
```diff
@@ -119,8 +111,23 @@ private static string GetFolderPathCoreWithoutValidation(SpecialFolder folder)
...
+                    // "$XDG_DATA_HOME defines the base directory relative to which user specific data files should be stored."
+                    // "If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used."
+                    string? data = GetEnvironmentVariable("XDG_DATA_HOME");
+                    if (string.IsNullOrEmpty(data) || data[0] != '/')
+                    {
+                        data = Path.Combine(home, ".local", "share");
+                    }
+                    return data;
+                case SpecialFolder.MyDocuments: // same value as Personal
+                    return ReadXdgDirectory(home, "XDG_DOCUMENTS_DIR", "Documents");
```
Comment (tmds) [inline src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:126]: There will be code that uses `MyDocuments` for `HOME`.
That code will break, and need to be updated to using `UserProfile` instead.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 2 above)
Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: > With no access to the previous location, implementing *settings migration* is also difficult.

The previous location was just `/home/.config` / `/home/.local/share` so that can be used for migration. Would need to be made clear for a release.

> is there a tangible improvement?

It follows Apple's Guidelines, which from what I know are necessary to fulfill in order to publish anything on their App Store.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 3 above)
Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:126]: With "that code" are you referring to code inside of .NET or user made programs?

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 2 above)
Comment (tmds) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: > It follows Apple's Guidelines, which from what I know are necessary to fulfill in order to publish anything on their App Store.

Without this change, .NET apps aren't accepted in the Apple App Store?

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 3 above)
Comment (tmds) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:126]: Both.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 2 above)
Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: Not as far as I know, but am also not a macOS developer so never dealt with that in any way.
[Based on](https://developer.apple.com/app-store/review/guidelines/#before-you-submit) :
> Check whether your app follows guidance in other documentation

I'd assume so at least.

**Comment by marek-safar** [comment]
> The mobile platforms are probably the most impacted here.

This should be fine for mobile but could break VS4M when moving to net7

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 2 above)
Comment (soc) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: Is there further information on using `~/Library` for `LOCAL_APP_DATA`?

I have used `~/Library/Application Support` for that as well (https://github.com/dirs-dev/directories-jvm#basedirectories) and haven't seen issues or user complaints.

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:98]: I based this mainly off [this StackOverflow post](https://stackoverflow.com/a/5084892) which in turn used this [rather old now Mac Java FAQ](https://developer.apple.com/library/archive/qa/qa1170/_index.html) and maybe the fact that the [iOS doc mentions](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW12) creating custom subdirectories.
I do however see people use `~/Library/Application Support` for [quite](https://doc.qt.io/qt-5/qstandardpaths.html) [a few](https://github.com/sindresorhus/env-paths) [other](https://github.com/adrg/xdg) [libs](https://github.com/sago007/PlatformFolders). So probably a good call to indeed change it to that instead.

Comment (soc) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs]: @tmds

> backwards compatibility

Having a good migration story is important, but we can't keep pretending that accreting technical debt instead of fixing things is fine. That's one of the things that brought computing into disrepute in the wider public.

Perhaps the behavior could be predicated on the .NET version targeted, such that only code compiled for the next version gets the correct folders?

>  It will break things, and the benefit is not clear to me (besides correctness).

Writing things to `$HOME` (e. g. `~/.dotnet`) also breaks things. You cannot expect `$HOME` to be writable. (Example: I make `$HOME` read-only on every machine under my control.)

Comment (tmds) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs]: For any breaking change you need to weigh the pros and cons. Calling it _technical debt_ doesn't change that.

Comment (soc) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs]: Yes, and?

Comment (soc) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs]: @tannergooding Agreed. If special folders are finally updated, let's make sure they are actually correct. Those who incorrectly relied on buggy behavior should make a suggestion on how to address their problems, not the people fixing things.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 1 above)
Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: Am not that familiar with the codebase, but my two cents are, that it probably would be better to put macOS into there as well, rename the file to `Environment.Apple`, and put some macOS specific checks in there (for i.e. `ApplicationData`, `Personal` etc.).

On another note, why does `Environment.iOS` hardcode the subdirectories, instead of invoking the respective NSLibrary paths?
For `SpecialFolder.Desktop` for example, `NSDocumentDirectory/Desktop` is used instead of `NSDesktopDirectory` which according to the Apple docs at least work on [all Apple OS](https://developer.apple.com/documentation/foundation/nssearchpathdirectory/nsdesktopdirectory). This is probably a thing for a separate PR, but was curious nonetheless.

(Also, [updated link](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.iOS.cs))

**Comment by danmoseley** [comment]
Perhaps at least part is low enough concern to go into .NET 7. And anything we are less sure about can go in first build of .NET 8 giving lots of time for anyone for any widespread breaks for mobile, VS4Mac etc to show up and be considered.

File: src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs [hunk 4]
```diff
@@ -393,7 +396,7 @@ public void GetSystemDirectory()
         [Theory]
         [PlatformSpecific(TestPlatforms.AnyUnix)]  // Tests OS-specific environment
         [InlineData(Environment.SpecialFolder.UserProfile, Environment.SpecialFolderOption.None)]
-        [InlineData(Environment.SpecialFolder.MyDocuments, Environment.SpecialFolderOption.None)] // MyDocuments == Personal
+        [InlineData(Environment.SpecialFolder.MyDocuments, Environment.SpecialFolderOption.DoNotVerify)] // MyDocuments == Personal
```
Comment (eerhardt) [inline src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:399]: Why the change to `DoNotVerify` here?

Comment (Miepee) [reply src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:399]: From what I remember, it was because the documents folder along with pictures/videos/music folder etc. don't exist on the test environment and thus would fail when being tested against.

Comment (eerhardt) [reply src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:399]: is that only for certain platforms? Which platforms does this folder not exist?

**Comment by danmoseley** [comment]
@eerhardt's going to sponsor this PR -- as @Miepee observed, anything about SpecialFolders tends to get stale and fail because we end up spinning on questions about breaking changes. Now we're starting a new release, we have the opportunity to take a little more risk if needed.

File: src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs (hunk 4 above)
Comment (Miepee) [reply src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:399]: I think only for Linux.
When I made the PR initially without that change, I couldn't run the tests locally, so I relied on the CI here to check whether tests succeed or not. They at first failed due to the folders not existing, so I changed it to DoNotVerify to be consistent with the other "media" folders and make the tests pass. It has been a few months, so don't remember all the details. I force pushed a lot, so comparing the commits is difficult + the original helix logs have become invalid now.

**Comment by eerhardt** [comment]
I've researched the changes here and I think this change is good.

The changes outside of the `SpecialFolder.MyDocuments/Personal` changes, while breaking, are correct and straight-forward. I don't think the those changes (they are all macOS-only) are controversial.

The change of `SpecialFolder.MyDocuments/Personal` on Unix however is a pretty big breaking change. I've searched for places using this value and sent PRs where I could to fix them. The most common theme is that people were using `SpecialFolder.Personal` when they really meant `SpecialFolder.UserProfile`. Because this appears to be a common case of confusion, I opened https://github.com/dotnet/runtime/issues/75563 to hide/deprecate `Personal`. Instead, developers should explicitly pick `MyDocuments` or `UserProfile` as appropriate.

I think this change is ready to go now that net8.0 is open. We should get it in as early as possible, so we can catch any breaks this will cause. I will also write up a breaking change document for it.

@Miepee - can you rebase this change on the latest `main`? That way we can get a fresh CI run, and we can merge it when the CI passes.

**Comment by Miepee** [comment]
Rebased! I assume the mono-llvm failure, and the tvOS cancel are unrelated?

**Comment by eerhardt** [comment]
Failure is #75606.

I'm going to leave this open one more day for people to express any objection feedback. But if I don't see any by tomorrow, I think we should merge this change and I'll write up the breaking change doc. (Note we can always revert it later, if we get enough feedback that this was the wrong decision/change.)

**Comment by Miepee** [comment]
No objections for 3 days now - @eerhardt wanna try merging?

**Comment by soc** [comment]
You could check against https://github.com/dirs-dev/directories-jvm#userdirectories and its https://github.com/dirs-dev/directories-jvm/blob/main/src/main/java/dev/dirs/UserDirectories.java.

As far as I have looked, the decisions seem to match, which is a good sign.

**Comment by Miepee** [comment]
Quickly checked MacOS, seems to match in both user dirs and base dirs.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:94]: I agree that if you're going to do breaking changes, do it right by using [NSSearchPathDirectory](https://developer.apple.com/documentation/foundation/nssearchpathdirectory) values as much as possible, like it's done for iOS here:

https://github.com/dotnet/runtime/blob/16cb35fda05165d679603c1c9c7ca5710e9bd3df/src/libraries/System.Private.CoreLib/src/System/Environment.iOS.cs#L84-L85

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:97]: As far as I can see, that'd require changing the method's return value from `string` to `string?` (as the `Interop.Sys.SearchPath` method can return null), which would be yet another API break. Is that a change worth having?

**Comment by Miepee** [comment]
I tried applying the suggestion to use `NSSearchPath`s for OSX, but seems like there's something that the tests don't like. Is that still the failure issue from above? Because from reading the logs, I'm not getting any smarter there.

**Comment by jkotas** [comment]
> EDIT: well, by the mention just below, i'm assuming another known infra-issue.

I do not think that this is a known issue.

I think you are hitting the unreachable assert here: https://github.com/dotnet/runtime/blob/7414af2a5f6d8d99efc27d3f5ef7a394e0b23c42/src/native/libs/System.Native/pal_searchpath.c#L10 . You need to change ifdefs around this native shim now that you have started using it in managed code.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs [hunk 5]
```diff
@@ -17,7 +20,7 @@ public static partial class Environment
         private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOption option)
         {
             // Get the path for the SpecialFolder
-            string path = GetFolderPathCoreWithoutValidation(folder);
+            string? path = GetFolderPathCoreWithoutValidation(folder);
             Debug.Assert(path != null);
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:24]: This doesn't seem valid. Why would `GetFolderPathCoreWithoutValidation` return a nullable string if it will never be `null`?

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:24]: `Interop.Sys.SearchPath()` returns a nullable string, which is used i.e. [here](https://github.com/dotnet/runtime/pull/68610/files#diff-a0d453801306c48ad8f563e3288251a7500e62cadee05aa27b155a6fc0864c04R100)
I also could've done something like `string? value = Interop.Sys.SearchPath(NSSearchPathDirectory.Foo); return value ?? String.Empty` for those situations, but thought this was cleaner.

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:24]: Then how is this assert valid?

**Comment by Miepee** [comment]
Thanks [jkotas](https://github.com/jkotas) for the hint, I think I figured out the place where to do that.

File: src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs (hunk 5 above)
Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs:24]: ...Oh I see what you mean now. I'll try fixing that.

Comment (Miepee) [reply src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs]: Added a fix, lmk if it's ok.

**Review by eerhardt — APPROVED** [review]
This LGTM. Thanks @Miepee for pushing this forward.

I'll write up the breaking change doc and post it here after I merge.

**Comment by RayKoopa** [comment]
Great work all! Once this is merged I'll begin implementing [my proposed additions to the `SpecialFolder` enum](https://github.com/dotnet/runtime/issues/70484) ASAP, as it's dependent on fixes being done here.

**Comment by eerhardt** [comment]
Breaking change document is here: https://github.com/dotnet/docs/issues/31423

**Comment by soc** [comment]
@eerhardt I think the line 

> `$XDG_DOCUMENTS_DIR`, if set. `$HOME/Documents`

needs to be clarified a bit, because it makes `$XDG_DOCUMENTS_DIR` look like an environment variable, not an entry from the `user-dirs.dirs` file.

In my documentation I use `XDG_DOCUMENTS_DIR` (without the dollar sign), maybe that's an option.

**Comment by Miepee** [comment]
It is both actually.
It first checks whether the environment variable is set, and if yes, chooses that.
If the env var is not set, it reads the value from `user-dirs.dirs`
And if it can't find it there, it chooses the fallback option.

But yeah, would probably be a good idea to clarify this behaviour for user XDG directories.

**Comment by soc** [comment]
> It is both actually.
> It first checks whether the environment variable is set, and if yes, chooses that.

Well, then that's absolutely wrong and should be corrected.

**Comment by steveisok** [comment]
This needs corrected on Android as `GetFolderPath_Unix_PersonalIsDocumentsAndUserProfile` fails on the rolling build with:

```
Assert.Equal() Failure
↓ (pos 58)
Expected: ···tensions.Tests/files/Documents
Actual:   ···tensions.Tests/files
↑ (pos 58)
```

@Miepee some of our configurations run post PR because we have limited capacity.  It is possible to run them on PR's, but someone on the dotnet team would have to trigger them on your behalf.  Not much you can do.  Thanks for your contribution!

**Comment by steveisok** [comment]
Cursory glance, this would need to be modified to conform with the breaking change:

https://github.com/dotnet/runtime/blob/b2af65a406bc28c25f0a25ed87b4e48c98f52fef/src/libraries/System.Private.CoreLib/src/System/Environment.Android.cs#L57-L59

**Comment by Miepee** [comment]
I ain't that familiar with Android's file structure, should I just change `Personal` to `$HOME/Documents` and keep `LocalApplicationData` to `$HOME`?

**Comment by steveisok** [comment]
> I ain't that familiar with Android's file structure, should I just change `Personal` to `$HOME/Documents` and keep `LocalApplicationData` to `$HOME`?

Yeah, I think that's fine.

**Comment by Miepee** [comment]
Opened #76250 to fix that.
